### PR TITLE
wxGraphicsContext: support double pen width

### PR DESCRIPTION
**Pull request by adrientetar**
Closes trac/17087.
Closes #464.

r? @vadz

### CODE REVIEW

**Comment by adrientetar** [comment]
Vadim, does the general approach here correspond to what you wanted? If so I'll clean up the patch / fix the compile errors.

File: include/wx/graphics.h [hunk 1]
```diff
@@ -34,7 +34,7 @@ enum wxInterpolationQuality
     // default interpolation
     wxINTERPOLATION_DEFAULT,
     // no interpolation
-    wxINTERPOLATION_NONE, 
+    wxINTERPOLATION_NONE,
```
Comment (vadz) [inline include/wx/graphics.h]: It would be nice to leave out the whitespace-only changes or make them in a separate commit if possible (here and elsewhere).

File: include/wx/graphics.h [hunk 2]
```diff
@@ -132,6 +132,49 @@ class WXDLLIMPEXP_CORE wxGraphicsObject : public wxObject
...
+
+    wxGraphicsPenInfo& Width(int width);
+
+    wxGraphicsPenInfo& Style(wxPenStyle style);
+    wxGraphicsPenInfo& Stipple(const wxBitmap& stipple);
+    wxGraphicsPenInfo& Dashes(int nb_dashes, const wxDash *dash);
+    wxGraphicsPenInfo& Join(wxPenJoin join);
+    wxGraphicsPenInfo& Cap(wxPenCap cap);
+
+    wxGraphicsPenInfo& WidthF(wxDouble widthF)
```
Comment (vadz) [inline include/wx/graphics.h]: I'm not sure if we use `FooF` naming convention anywhere else and it seems not very clear to me, especially for a function taking `double`.

In `wx/gdicmn.h` we have `wxRealPoint` etc as counterparts to `wxPoint`, so I think it would be more consistent to use `RealWidth()` here, although, admittedly, this does look a bit confusing (is the other width unreal then?). Maybe `WidthReal()` would be more clear, even though it sounds undoubtedly worse.

File: include/wx/pen.h [hunk 3]
```diff
@@ -59,6 +59,86 @@ enum wxPenCap
...
+    void Init()
+    {
+        m_stipple = wxNullBitmap;
+        m_nb_dashes = 0;
+        m_dash = NULL;
+        m_join = wxJOIN_ROUND;
+        m_cap = wxCAP_ROUND;
+    }
+
+    wxColour& m_colour;
```
Comment (vadz) [inline include/wx/pen.h]: This should be an object, why is it a reference, _especially_ a non-const one? This is very, very wrong: you could change a colour by creating a pen using it and then changing the colour of this pen which doesn't make sense at all!

File: include/wx/pen.h [hunk 4]
```diff
@@ -59,6 +59,86 @@ enum wxPenCap
...
 
+// ----------------------------------------------------------------------------
+// wxPenInfo describes a wxPen
+// ----------------------------------------------------------------------------
+
+class wxPenInfo
+{
+public:
+    wxPenInfo() :
+        m_colour(wxNullColour)
```
Comment (vadz) [inline include/wx/pen.h]: This is unnecessary.

File: include/wx/pen.h [hunk 5]
```diff
@@ -59,6 +59,86 @@ enum wxPenCap
...
+    {
+        Init();
+
+        m_width = 1;
+        m_style = wxPENSTYLE_SOLID;
+    }
+
+    explicit wxPenInfo(const wxColour& colour, int width = 1, wxPenStyle style = wxPENSTYLE_SOLID) :
+        m_colour((wxColour&) colour)
```
Comment (vadz) [inline include/wx/pen.h]: There should be no cast here.

File: include/wx/pen.h [hunk 6]
```diff
@@ -59,6 +59,86 @@ enum wxPenCap
...
+    wxPenInfo() :
+        m_colour(wxNullColour)
+    {
+        Init();
+
+        m_width = 1;
+        m_style = wxPENSTYLE_SOLID;
+    }
+
+    explicit wxPenInfo(const wxColour& colour, int width = 1, wxPenStyle style = wxPENSTYLE_SOLID) :
```
Comment (vadz) [inline include/wx/pen.h]: In principle, by adding `= wxColour()` to the first argument, this ctor could replace the default ctor too. Notice that, unlike in `wxPen` case, there is no advantage in having a separate implementation for the default ctor (for `wxPen` there is, as the default ctor doesn't create any pen object, while the other one does, but this is irrelevant here).

File: include/wx/pen.h [hunk 7]
```diff
@@ -59,6 +59,86 @@ enum wxPenCap
...
+        { m_nb_dashes = nb_dashes; m_dash = (wxDash *)dash; return *this; }
+    wxPenInfo& Join(wxPenJoin join)
+        { m_join = join; return *this; }
+    wxPenInfo& Cap(wxPenCap cap)
+        { m_cap = cap; return *this; }
+
+    // Accessors are mostly meant to be used by wxPen itself.
+
+    wxColour GetColour() const { return m_colour; }
+    wxBitmap* GetStipple() { return &m_stipple; }
```
Comment (vadz) [inline include/wx/pen.h]: Why pointer? Could, and IMHO should, return the object.

File: include/wx/pen.h [hunk 8]
```diff
@@ -59,6 +59,86 @@ enum wxPenCap
...
+    int GetDashes(wxDash **ptr) const { *ptr = m_dash; return m_nb_dashes; }
+
+    // Convenience
+
+    bool IsTransparent() const { return m_style == wxPENSTYLE_TRANSPARENT; }
+
+private:
+    void Init()
+    {
+        m_stipple = wxNullBitmap;
```
Comment (vadz) [inline include/wx/pen.h]: This is unnecessary.

File: interface/wx/graphics.h [hunk 9]
```diff
@@ -1527,6 +1532,46 @@ class wxGraphicsFont : public wxGraphicsObject
...
+ */
+class wxGraphicsPenInfo : public wxPenInfo
+{
+public:
+
+    wxGraphicsPenInfo();
+
+    explicit wxGraphicsPenInfo(const wxColour& colour, double widthF = 1.0, wxPenStyle style = wxPENSTYLE_SOLID);
+
+    wxGraphicsPenInfo& ();
```
Comment (vadz) [inline interface/wx/graphics.h]: There is something missing here.

File: interface/wx/pen.h [hunk 10]
```diff
@@ -100,6 +100,43 @@ enum wxPenCap
...
 
 
+/**
+    @class wxPenInfo
+
+    This class is a helper used for wxPen creation using named parameter
+    idiom: it allows to specify various wxPen attributes using the chained
+    calls to its clearly named methods instead of passing them in the fixed
+    order to wxPen constructors.
+
```
Comment (vadz) [inline interface/wx/pen.h]: I think an example would be very nice to have here.

File: interface/wx/pen.h [hunk 11]
```diff
@@ -100,6 +100,43 @@ enum wxPenCap
...
+    @since 3.1.0
```
Comment (vadz) [inline interface/wx/pen.h]: 3.1.1

File: src/generic/graphicc.cpp [hunk 12]
```diff
@@ -847,14 +872,14 @@ wxCairoPenData::wxCairoPenData( wxGraphicsRenderer* renderer, const wxPen &pen )
     case wxPENSTYLE_STIPPLE :
     case wxPENSTYLE_STIPPLE_MASK :
     case wxPENSTYLE_STIPPLE_MASK_OPAQUE :
-        InitStipple(pen.GetStipple());
+        InitStipple(((wxGraphicsPenInfo&) info).GetStipple());
```
Comment (vadz) [inline src/generic/graphicc.cpp]: There should be no need to cast here.

File: src/gtk/pen.cpp [hunk 13]
```diff
@@ -103,6 +103,19 @@ wxPen::wxPen(const wxColour& colour, int width, int style)
...
 
+wxPen::wxPen(const wxPenInfo& info)
+{
+    m_refData = new wxPenRefData();
+    M_PENDATA->m_colour = info.GetColour();
+    M_PENDATA->m_width = info.GetWidth();
+    M_PENDATA->m_style = info.GetStyle();
+    M_PENDATA->m_joinStyle = info.GetJoin();
+    M_PENDATA->m_capStyle = info.GetCap();
+    wxDash *dashes;
```
Comment (vadz) [inline src/gtk/pen.cpp]: This temp var seems unnecessary, could just do `info.GetDashes(&M_PENDATA->m_dash)` directly.

File: src/gtk1/pen.cpp [hunk 14]
```diff
@@ -103,6 +103,20 @@ wxPen::wxPen(const wxColour& colour, int width, int style)
...
+wxPen::wxPen(const wxPenInfo& info)
+{
+    m_refData = new wxPenRefData();
+
+    M_PENDATA->m_colour = info.GetColour();
+    M_PENDATA->m_width = info.GetWidth();
+    M_PENDATA->m_style = info.GetStyle();
+    M_PENDATA->m_joinStyle = info.GetJoin();
+    M_PENDATA->m_capStyle = info.GetCap();
+    wxDash *dashes;
```
Comment (vadz) [inline src/gtk1/pen.cpp]: Same as above.

File: src/generic/graphicc.cpp [hunk 15]
```diff
@@ -735,12 +737,35 @@ void wxCairoPenData::Init()
 wxCairoPenData::wxCairoPenData( wxGraphicsRenderer* renderer, const wxPen &pen )
     : wxCairoPenBrushBaseData(renderer, pen.GetColour(), pen.IsTransparent())
 {
+    wxDash *dashes;
+    int nb_dashes = pen.GetDashes(&dashes);
+    InitFromPenInfo(wxGraphicsPenInfo()
```
Comment (vadz) [inline src/generic/graphicc.cpp]: It looks like we're doing this (constructing `wxGraphicsPenInfo` from an existing `wxPen`) in a couple of places, so I'd provide a `static wxGraphicsPenInfo CreateFromPen(const wxPen& pen)` method doing it and use it all of these places.

File: src/msw/graphicsd2d.cpp [hunk 16]
```diff
@@ -2474,9 +2478,9 @@ class wxD2DPenData : public wxGraphicsObjectRefData, public wxD2DManagedGraphics
     }
 
 private:
-    // We store the source pen for later when we need to recreate the
+    // We store the source info for later when we need to recreate the
     // device-dependent resources.
-    const wxPen m_sourcePen;
+    const wxGraphicsPenInfo& m_sourceInfo;
```
Comment (vadz) [inline src/msw/graphicsd2d.cpp]: This should be an object too, you seem to be using a dangling reference, i.e. this will result in an all but guaranteed runtime crash!

File: include/wx/graphics.h [hunk 17]
```diff
@@ -132,6 +132,49 @@ class WXDLLIMPEXP_CORE wxGraphicsObject : public wxObject
...
+    explicit wxGraphicsPenInfo(const wxColour& colour, double widthF = 1.0, wxPenStyle style = wxPENSTYLE_SOLID)
+    : wxPenInfo(colour, 0, style)
+    {
+        m_widthF = widthF;
+    }
+
+    // Setters for the various attributes. All of them return the object itself
+    // so that the calls to them could be chained.
+
+    wxGraphicsPenInfo& Colour(const wxColour& colour);
```
Comment (vadz) [inline include/wx/graphics.h]: Where is this and other methods implemented? I don't see them...

In principle, we could omit the methods present at `wxPenInfo` level. This would mean that `WidthReal()` (or whatever we end up calling it) would have to be called first, but maybe it's not such a huge limitation?

We could also use some template trickery to avoid having to provide all these methods here:
```cpp
template <class T>
class wxPenInfoBase
{
public:
   T& Colour(const wxColour& colour) { m_colour = colour; return static_cast<T&>(*this); }
   ... same thing for all the other methods ...
private:
   wxColour m_colour;
};

class wxPenInfo : public wxPenInfoBase<wxPenInfo>
{
   using wxPenInfoBase<wxPenInfo>::wxPenInfoBase;
};

class wxGraphicsPenInfo : public wxPenInfoBase<wxGraphicsPenInfo>
{
public:
   using wxPenInfoBase<wxGraphicsPenInfo>::wxPenInfoBase;

    wxGraphicsPenInfo& WidthReal(double w) { ... }
};
```

This would make it impossible to construct `wxPen` from `wxGraphicsPenInfo` too which, for me, is a bonus (because `wxPen` doesn't support floating point width).

**Review by vadz — CHANGES_REQUESTED** [review]
After thinking more about this, I believe that the idea with `wxPenInfoBase` is actually better.

But the really critical problem is the wrong use of references in several places.

**Comment by vadz** [comment]
Thanks for the PR! I've had a look at it and unfortunately there are some real problems with it currently. I also think the improvement with `wxPenInfoBase` is worth it, but if you prefer not to do it like this, it would be acceptable too -- but then you'd need to actually implement `wxGraphicsPenInfo` methods which don't seem to be implemented now. I also think we need to change the name of the new accessor from `WidthF`.

Thanks again!

File: include/wx/graphics.h (hunk 2 above)
Comment (adrientetar) [inline include/wx/graphics.h]: Maybe I could just make width a double in wxGraphicsPenInfo? It seems all underlying implementations just store a double anyway.

Comment (vadz) [inline include/wx/graphics.h]: `wxDouble` is just `double` anyhow.

Comment (adrientetar) [inline include/wx/graphics.h]: I meant, there's no need to store int width at all in wxGraphicsPenInfo.

File: include/wx/pen.h (hunk 8 above)
Comment (adrientetar) [inline include/wx/pen.h]: I should leave m_stipple uninitialized here?

Comment (vadz) [inline include/wx/pen.h]: Well, yes, what's the point of reinitializing an object which already is properly initialized by its default ctor to the same value?

File: include/wx/peninfo.h [hunk 18]
```diff
@@ -0,0 +1,145 @@
+/////////////////////////////////////////////////////////////////////////////
+// Name:        wx/peninfo.h
```
Comment (vadz) [inline include/wx/peninfo.h]: The new header needs to be added to `build/files` and `upmake` and `bakefile_gen` rerun, otherwise it won't be installed by `make install`.

File: include/wx/peninfo.h [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+    int GetDashes(wxDash **ptr) const { *ptr = m_dash; return m_nb_dashes; }
+
+    // Convenience
+
+    bool IsTransparent() const { return m_style == wxPENSTYLE_TRANSPARENT; }
+
+private:
+    void Init()
+    {
+        /* m_stipple skipped */
```
Comment (vadz) [inline include/wx/peninfo.h]: Even this comment is superfluous, there is nothing remarkable about not initializing an already initialized object member.

File: include/wx/peninfo.h [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+    wxDash* m_dash;
+};
+
+// ----------------------------------------------------------------------------
+// wxPenInfo describes a wxPen
+// ----------------------------------------------------------------------------
+
+class wxPenInfo : public wxPenInfoBase<wxPenInfo>
+{
+    using wxPenInfoBase<wxPenInfo>::wxPenInfoBase;
```
Comment (vadz) [inline include/wx/peninfo.h]: This doesn't actually change anything for the compiler (the ctor is still public because it's public in the base class), but I'd put this in the public section of this class too, to avoid any confusion.

But, actually, wait: this is not needed at all because there are no ctors in the base class now that `wxGraphicsPenInfo` doesn't accept int width any more. So either the ctor from `int width` should be moved to the base class or this line should just be removed completely.

File: include/wx/peninfo.h [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+
+        m_colour = colour;
+        m_width = width;
+        m_style = style;
+    }
+
+    // Setters
+
+    wxPenInfo& Width(int width)
+        { m_width = width; return static_cast<wxPenInfo&>(*this); }
```
Comment (vadz) [inline include/wx/peninfo.h]: There is no need for the cast here AFAICS.

**Review by vadz — COMMENTED** [review]
Thanks, this looks almost perfect now.

I'd still like to see an example in the docs and, now that I think about it, it would be also very useful to use the new API in `samples/drawing/drawing.cpp` to check that it works correctly. TIA!

File: include/wx/peninfo.h (hunk 18 above)
Comment (adrientetar) [inline include/wx/peninfo.h]: I get errors in peninfo.h that all the wxPen constants aren't defined (wxPenStyle, etc.), even though I added an include for pen.h. What is the proper way of including pen.h here?

Comment (vadz) [inline include/wx/peninfo.h]: I'm not sure what's going on here, but could it be that `wx/peninfo.h` is included from `wx/pen.h` before these constants are defined in it?

In any case, it is definitely a bad idea to have inclusions in both directions between these files. So either `wxPenInfo` itself should just to into the existing `wx/pen.h` or the styles need to be moved into `wx/peninfo.h` too.

**Comment by adrientetar** [comment]
The buildbot failure on OSX seems unrelated to this PR; otherwise I think this is getting ready for landing.

**Comment by adrientetar** [comment]
Can this be merged?

**Comment by vadz** [comment]
Sorry for the delay, I'll try to get back to this a.s.a.p.

**Comment by adrientetar** [comment]
Do you know when you'll release 3.1.1?

**Comment by vadz** [comment]
Sorry, I don't have any exact date in mind. I do currently intend to apply (or otherwise deal with) all the PRs using 3.1.1 milestone until then.

**Comment by adrientetar** [comment]
I really need this for my project, is it possible to merge it in and release? Or maybe there's another way to do this: I draw graphics on a canvas with zoom functionality so I scale my drawing according to the zoom and basically divide pen width by the zoom scale so that pen width is constant regardless of zoom level.

Is it possible to do that with the current wxwidgets?

Currently my pen widths get rounded to int so things look bad especially at higher zoom levels:

zoom

becomes

zoom2

I'm getting push to get back to developing my font editor software but I want to switch to wxwidgets and right now I'm stuck with an unfinished port because of this issue.

**Comment by vadz** [comment]
Sorry for the delay, I'll do my best to get back to this this week-end.

**Comment by vadz** [comment]
I've changed a few things and created PR #553 with the combined changes (your commits are different because I cherry-picked them ignoring whitespace-only changes, but without changing them otherwise). If the CI builds pass, I'll merge it tomorrow.

**Comment by vadz** [comment]
Thanks for all your work on this and sorry once again for the delay.

**Comment by adrientetar** [comment]
Thanks, Vadim. I know what it's like to maintain open-source projects.
